### ISSUE TITLE
Ctrl+scroll Tool Selection Issues on Mac

### ISSUE THREAD

**Issue by Lapask** [maruohon/litematica#36]
It seems as though control+scrolling is finicky on Mac.

**Comment by maruohon** [comment]
Yeah I've heard the same report from others recently. I'll try to change some of the input handling methods to use doubles instead of casting to ints, which should hopefully fix the issue of "weird" small scroll inputs getting lost currently.

Unfortunately I think that change may need to wait for the malilib 0.10.0 release and the corresponding round of updates to all the dependant mods, as it will break compatibility with the current/old versions.

**Comment by maruohon** [comment]
Could you try these builds, if they fix your issue?

https://masa.dy.fi/tmp/minecraft/mods/malilib/malilib-rift-1.13.2-0.9.5.jar
https://masa.dy.fi/tmp/minecraft/mods/litematica/litematica-rift-1.13.2-0.0.0-dev.20190519.002745.jar

**Comment by spencerjw** [comment]
> Could you try these builds, if they fix your issue?
> 
> https://masa.dy.fi/tmp/minecraft/mods/malilib/malilib-rift-1.13.2-0.9.5.jar
> https://masa.dy.fi/tmp/minecraft/mods/litematica/litematica-rift-1.13.2-0.0.0-dev.20190519.002745.jar

Even the latest version of Litematica (litematica-fabric-1.16.4-0.0.0-dev.20201103.184101.jar) and MaLiLib (malilib-fabric-1.16.4-0.10.0-dev.21+arne.2.jar) downloaded yesterday (Dec. 23, 2020) are not working for Ctrl+Scrol on a Mac. Suggestions?

**Comment by maruohon** [comment]
@spencerjw Can you test this build: https://masa.dy.fi/mcmods/litematica/litematica-fabric-1.16.4-0.0.0-dev.20201225.203013.jar

There is a new option in the Generic category: `debugLogging`.
Enable that and then try the normal Ctrl + scrolling while holding the tool, and you should see some debug prints in the game console, like this:
```
[20:28:52] [main/INFO]: Mouse scroll: x: 416, y; 191, wheel: 1.0
[20:28:53] [main/INFO]: Mouse scroll: x: 416, y; 191, wheel: -1.0
```
So I'm curious what the wheel values are (on a Mac?).

**Comment by spencerjw** [comment]
@maruohon I gave that new file a try and found the debugLogging option and set it to true. But When running the ctrl+Scroll I'm not seeing anything printing to the game console window.

**Comment by itakashiraiya** [comment]
@maruohon So, my scroll wheel is broken, i can still click it, but i cant scroll, is there any way to change the inout for the mode change to something without it?

**Comment by maruohon** [comment]
@itakashiraiya Not in the current version. The handling happens from the mouse input handling, when scroll events are detected, and the modifier hotkey being pressed is checked there. I recently added mouse scrolls to the keybind system in malilib in the main dev version, so in future releases it would be possible to add the full `modifier + scroll` as a hotkey. If I add those, they will likely be new hotkeys, since I don't really want to get rid of the current modifier + hard coded scroll either.

So in the current version you would just have to use the non-scroll input options, such as using the tool mode button in the bottom left of the Litematica menu, and I believe there are also separate nudge hotkeys for moving selections around, other than the `selectionNudgeModifier` + scroll (or at least there are in 1.12.2).

**Comment by zdwolfe** [comment]
I'm having this issue as well on Mac OSX, Litematica v0.24.3, MaLiLib v0.26.3. I enabled the ``mouseScrollDebug`` which prints [logs](https://gist.github.com/zdwolfe/b1184ef3d89a84344e309a4138d082ce).

I think adding some configurable hotkeys for cycling the tool mode would be a great workaround here. I see that ``toolModeCycleForward`` and ``toolModeCycleBackward`` are already [defined](https://github.com/maruohon/litematica/blob/271833c7a61d9a9785e3568686cddfef6e22276f/src/main/java/litematica/config/Hotkeys.java#L70-L71), and [registered](https://github.com/maruohon/litematica/blob/271833c7a61d9a9785e3568686cddfef6e22276f/src/main/java/litematica/config/HotkeyCallbackMisc.java#L403-L407) but they are not visible in the configuration menu. Adding them directly to ``.minecraft/config/litematica.json`` doesn't appear to work, and they get blown away when the config is saved.

Looking around, it seems like this is related: https://github.com/maruohon/litematica/issues/671

**Comment by maruohon** [comment]
@zdwolfe The code you linked is from the main development 1.12.2 branches, where those hotkeys have existed for a long time. That code still hasn't been merged to any of the 1.13+ mod versions.

As for the debug log, I don't remember the code without going to look, but if it says "amount: -1.0", then why is it not doing anything...?

You shouldn't actually *need* to use the scroll wheel for anything, but ofc not being able to use it will make some things a bit slower to do. You can change the tool mode in the bottom left of the main menu, and you can adjust the placement and area selection position via the placement config menu and the area editor respectively. (btw does scrolling over the +/- adjustment buttons in the menus work either on your system?)

**Comment by zdwolfe** [comment]
@maruohon Thanks for your reply! 

> The code you linked is from the main development 1.12.2 branches, where those hotkeys have existed for a long time. That code still hasn't been merged to any of the 1.13+ mod versions.

Ah, I see! Thank you.

---

> You can change the tool mode in the bottom left of the main menu, and you can adjust the placement and area selection position via the placement config menu and the area editor respectively. 

Oh, nice, this is perfect. I didn't see that somehow. For anyone that finds this issue in the future, just hit M and toggle this button.

<img width="712" height="214" alt="Image" src="https://github.com/user-attachments/assets/9b06fa65-10a8-49e9-8fec-8c9524f28383" />
